apollo-cache-interceptor/src/commonMain/kotlin/com/apollographql/apollo/interceptor/cache/ApolloCacheInterceptor.kt

```diff
@@ -0,0 +1,95 @@
+package com.apollographql.apollo.interceptor.cache
+
+import com.apollographql.apollo.api.ApolloExperimental
+import com.apollographql.apollo.api.Operation
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.api.Response.Companion.builder
+import com.apollographql.apollo.api.ResponseField
+import com.apollographql.apollo.api.internal.NoOpResolveDelegate
+import com.apollographql.apollo.cache.CacheHeaders
+import com.apollographql.apollo.cache.normalized.CacheKey
+import com.apollographql.apollo.cache.normalized.CacheKeyResolver
+import com.apollographql.apollo.cache.normalized.Record
+import com.apollographql.apollo.cache.normalized.internal.CacheFieldValueResolver
+import com.apollographql.apollo.cache.normalized.internal.CacheKeyBuilder
+import com.apollographql.apollo.cache.normalized.internal.ReadableStore
+import com.apollographql.apollo.cache.normalized.internal.RealCacheKeyBuilder
+import com.apollographql.apollo.cache.normalized.internal.ResponseNormalizer
+import com.apollographql.apollo.cache.normalized.simple.SimpleNormalizedCache
+import com.apollographql.apollo.interceptor.ApolloInterceptorChain
+import com.apollographql.apollo.interceptor.ApolloRequest
+import com.apollographql.apollo.interceptor.ApolloRequestInterceptor
+import com.apollographql.apollo.interceptor.ApolloResponse
+import com.apollographql.apollo.internal.response.RealResponseReader
+import com.apollographql.apollo.internal.response.RealResponseWriter
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.flow
+
+@ApolloExperimental
+class ApolloCacheInterceptor : ApolloRequestInterceptor {
```

Comment:
should we mark as internal? I was thinking back in a day that cache plugin will take existing `ApolloClient` and decorate it with proper execution context and setup interceptors, so user won't know existence of interceptor as it's our internal implementation abstraction.

Reply:
Also excellent question. I'd love to find a nicer API. Did you have something in mind already? 

Should we introduce a `ApolloClientBuilder` to take care of the wiring of all the interceptors? In all cases, I think this should stay public as ultimately, we also want users to be able to plug their own interceptors and modify the list as they please?

Reply:
@sav007 I ended up adding `ApolloClient.Builder`, one option would have been to make `ApolloClient` a data class and copying it around. Using a Builder adds some boilerplate code but also has a few advantages:

* I didn't really liked the idea of making `ApolloClient` a data class
* It makes it clear what is mutable and what's not
* It puts the arguments verification logic in a single place